cpupower/cpupower.py

```diff
@@ -0,0 +1,181 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+#
+# Copyright: 2016 IBM
+# Author: Pavithra D P <pavithra@linux.vnet.ibm.com>
+
+import random
+from avocado import Test
+from avocado import main
+from avocado.utils import process
+
+
+class cpupower(Test):
+
+    """
+    Testing cpupower command
+    """
+
+    def test(self):
+        self.nfail = 0
+        self.i = 0
+        self.log.info("Get the initial values from the system")
+        (initial_freq, initial_governor) = self.get_initial_freq()
+        governors = self.get_list_governors()
+        (min, max, cur) = self.get_freeq()
+        for gov in governors:
+            if (gov == "performance"):
+                self.log.info("Check for performance governor")
+                self.check_performance_governor(gov, max)
+            if (gov == "powersave"):
+                self.log.info("Check for powersave governor")
+                self.check_powersave_governor(gov, min)
+            if (gov == "userspace"):
+                self.log.info("Check for userspace governor")
+                self.check_userspace_governor(gov)
+            if (gov == "ondemand" or "conservative" or "schedutils"):
+                self.log.info("Dynamic governors,need manual verification")
+        self.log.info("Set the final values on the system")
+        self.final_freq(initial_freq, initial_governor)
+        if self.nfail != 0:
+            self.fail("Test failed with errors")
+            self.log.info("The value of error count %s" % (self.nfail))
+        else:
+            self.log.info("Test passed successfully")
+
+    def get_initial_freq(self):
+        """
+        To get the initial frequency and governor set on system
+        """
+        cmd = "cat /sys/devices/system/cpu/cpu%s/cpufreq/scaling_cur_freq" % (self.i)
+        initial_freq = process.system_output(cmd)
+        cmd1 = "cat /sys/devices/system/cpu/cpu%s/cpufreq/scaling_governor" % (self.i)
+        initial_governor = process.system_output(cmd1)
+        return (initial_freq, initial_governor)
+
+    def get_list_governors(self):
+        """
+        Get the list of governors available on the system
+        """
+        cmd_governor_list = "cat /sys/devices/system/cpu/cpu%s/cpufreq/scaling_available_governors" % (self.i)
+        output = process.system_output(cmd_governor_list)
+        governors = output.split()
+        return governors
+
+    def get_freeq(self):
+        """
+        Get the cur,min,max freequency on the system
+        """
+        cmd1 = "cat /sys/devices/system/cpu/cpu%s/cpufreq/scaling_min_freq" % (self.i)
+        cmd2 = "cat /sys/devices/system/cpu/cpu%s/cpufreq/scaling_max_freq" % (self.i)
+        cmd3 = "cat /sys/devices/system/cpu/cpu%s/cpufreq/scaling_cur_freq" % (self.i)
```

Comment:
To avoid redundancy, you can have a method to form it. As it is almost similar every where,
```

def cpu_freq_path(self, cpu_num, path=None):
    cmd = "cat /sys/devices/system/cpu/cpu%s/cpufreq" % cpu_num
    if path:
        cmd = os.path.join(cmd, path)
    return cmd
```